### PR TITLE
Adding custom cmdlets for RSC configuration

### PR DESCRIPTION
**Pull request by subray2014**
Added cmdlets:
Get-MgBetaTeamRscConfiguration

Get-MgBetaChatRscConfiguration

Set-MgBetaTeamRscConfiguration -State {Disabled | EnabledForPreApprovedAppsOnly | EnabledForAllApps}

Set-MgBetaTeamRscConfiguration -State {Disabled | EnabledForPreApprovedAppsOnly | EnabledForAllApps}

﻿

Fixes #

### Changes proposed in this pull request

-
-
-

### Other links

-
-
-

### CODE REVIEW

File: src/Teams/beta/custom/RscConfigurationConverter.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                Id = "TeamResourceSpecificPermissionConfiguration",
+                ScopeType = MicrosoftGraphRscConfigurationScopeType.Team
+            };
+
+            MGTeamsInternalTenantConsentSettings groupConsentSettings = tenantConsentSettingCollection.Value?.FirstOrDefault(
+                v => string.Equals(v.TemplateId, RscConfigurationConverter.GroupConsentSettingsTemplateId, StringComparison.OrdinalIgnoreCase));
+
+            if (groupConsentSettings == null)
+            {
+                this.LogVerbose("Group Consent settings were not found.", eventListener);
```
Comment (AkJo) [inline src/Teams/beta/custom/RscConfigurationConverter.cs:121]: If null, then check the authorization policy to compute the semantics of the null by looking for user consent settings Legacy Default ON.
image

image

File: src/Teams/beta/custom/SetMgBetaTeamRscConfiguration_Update.cs [hunk 2]
```diff
@@ -0,0 +1,447 @@
+﻿namespace Microsoft.Graph.Beta.PowerShell.Cmdlets
+{
+    using Microsoft.Graph.Beta.PowerShell.Models;
+    using Microsoft.Graph.Beta.PowerShell.Models.TeamsInternal;
+    using Microsoft.Graph.Beta.PowerShell.TeamsInternal;
+    using System;
```
Comment (AkJo) [inline src/Teams/beta/custom/SetMgBetaTeamRscConfiguration_Update.cs:6]: Does this need sorting?

File: src/Teams/beta/custom/SetMgBetaTeamRscConfiguration_Update.cs [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+            if (groupConsentSettings == null)
+            {
+                // Settings need to be initialized. The default values are sourced from Azure defaults.
+                await this.Client.CreateGroupConsentSettings(
+                    new MGTeamsInternalTenantConsentSettingValue[]
+                    {
+                        new MGTeamsInternalTenantConsentSettingValue(RscConfigurationConverter.EnableGroupSpecificConsentKey, isGroupSpecificConsentEnabled.ToString()),
+                        new MGTeamsInternalTenantConsentSettingValue("BlockUserConsentForRiskyApps", true.ToString()),
+                        new MGTeamsInternalTenantConsentSettingValue("EnableAdminConsentRequests", false.ToString()),
+                        new MGTeamsInternalTenantConsentSettingValue("ConstrainGroupSpecificConsentToMembersOfGroupId", "")
```
Comment (AkJo) [inline src/Teams/beta/custom/SetMgBetaTeamRscConfiguration_Update.cs:397]: Let's fetch these values from DirectorySettingsTemplates > consentPolicySettings

image

File: src/Teams/beta/custom/RscConfigurationConverter.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        /// <param name="rscPermissionCollection">RSC permisssion collection.</param>
+        internal RscConfigurationConverter()
+        {
+        }
+
+        /// <summary>
+        /// Convert the given set of conditions to a permission grant preapproval policy.
+        /// </summary>
+        /// <param name="resourceSpecificApplicationPermissionsAllowedForChats">RSC permissions allowed for chats.</param>
+        /// <param name="teamLevelSensitivityLabelSelectionMode">Team level sensitivity label selection mode.</param>
```
Comment (AkJo) [inline src/Teams/beta/custom/RscConfigurationConverter.cs:34]: Comments needs to be updated.

File: src/Teams/beta/custom/SetMgBetaChatRscConfiguration_Update.cs [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+                                Pipeline);
+
+                            WriteVerbose($"Updated permission grant policies assigned to default user role: '{string.Join(", ", updatedPermissionGrantPolicies)}'.");
+                        }
+
+                        if (((Microsoft.Graph.Beta.PowerShell.Runtime.IEventListener)this).Token.IsCancellationRequested) { return; }
+                    }
+                    else if (this.State == MicrosoftGraphRscConfigurationState.EnabledForPreApprovedAppsOnly)
+                    {
+                        // Disable chat RSC Teams Setting.
```
Comment (AkJo) [inline src/Teams/beta/custom/SetMgBetaChatRscConfiguration_Update.cs:280]: Let's flip the order. First add preapproval policy and then mark the global setting as disabled.

File: src/Teams/beta/custom/SetMgBetaChatRscConfiguration_Update.cs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+                                this,
+                                Pipeline);
+
+                            WriteVerbose($"Updated permission grant policies assigned to default user role: '{string.Join(", ", updatedPermissionGrantPolicies)}'.");
+                        }
+
+                        if (((Microsoft.Graph.Beta.PowerShell.Runtime.IEventListener)this).Token.IsCancellationRequested) { return; }
+
+                        // Enable chat RSC Teams Setting.
+                        await this.Client.UpdateTeamsAppSettings(
```
Comment (AkJo) [inline src/Teams/beta/custom/SetMgBetaChatRscConfiguration_Update.cs:325]: Let's flip the order here. 

1. Enable global switch first.
2. Remove preapproval policy last.

File: src/Teams/beta/custom/SetMgBetaTeamRscConfiguration_Update.cs [hunk 7]
```diff
@@ -0,0 +1,447 @@
+﻿namespace Microsoft.Graph.Beta.PowerShell.Cmdlets
+{
+    using Microsoft.Graph.Beta.PowerShell.Models;
```
Comment (AkJo) [inline src/Teams/beta/custom/SetMgBetaTeamRscConfiguration_Update.cs:3]: Let's update public contract to:

Get-MgBetaTeamRscConfiguration

Get-MgBetaChatRscConfiguration

Set-MgBetaTeamRscConfiguration -State {DisabledForAllApps | EnabledForPreApprovedAppsOnly | EnabledForAllApps}

Set-MgBetaTeamChatConfiguration -State {DisabledForAllApps | EnabledForPreApprovedAppsOnly | EnabledForAllApps

**Review by peombwa — APPROVED** [review]

**Pull request merged**